python/tvm/relay/ir_builder.py

```diff
@@ -0,0 +1,299 @@
+    elif isinstance(arg, float):
+        return tvm.nd.array(arg, ctxt)
+    elif isinstance(arg, bool):
+        return tvm.nd.array(np.array(arg, dtype='float32'), ctxt)
+    elif isinstance(arg, np.ndarray):
+        return tvm.nd.array(arg, ctxt)
+    elif isinstance(arg, tvm.ndarray.NDArray):
+        return arg
+    else:
+        # raise Exception(f"can't convert {type(arg)} to a Relay AST")
+        raise Exception(f"unsupported argument type {type(arg)}")
+
+
+def into_ast(arg: Any, ctxt=tvm.cpu(0)) -> Expr:
+    if isinstance(arg, Expr):
+        return arg
+    elif isinstance(arg, tuple):
+        raise Exception("..")
+    elif isinstance(arg, PartialFunc):
+        return arg.to_func()
+    else:
+        value = convert(arg, ctxt)
+        return Constant(value)
+
+
+class WithScope(object):
+    """A wrapper for builder methods which introduce scoping."""
+
+    def __init__(self, enter_value, exit_cb):
+        self._enter_value = enter_value
+        self._exit_cb = exit_cb
+
+    def __enter__(self):
+        return self._enter_value
+
+    def __exit__(self, ptype, value, trace):
+        if value:
+            raise value
+        else:
+            self._exit_cb()
+
+
+class PartialFunc():
+    """A wrapper around functions while they are being built."""
+
+    def __init__(self, params, ret_type, body, type_params):
+        self.params = params
+        self.ret_type = ret_type
+        self.body = body
+        self.type_params = type_params
+
+    def param_ids(self):
+        return [p.var for p in self.params]
+
+    def to_func(self):
+        return Function(
+            self.params,
+            self.ret_type,
+            self.body,
+            self.type_params)
+
+#pylint: disable=invalid-name
+
+
+def _mk_let(bindings, ret_value):
+    let_expr = ret_value
+    for var, (value, ty) in reversed(list(bindings.items())):
+        let_expr = Let(var, value, let_expr, ty)
+
+    return let_expr
+
+
+class IRBuilder():
+    """The IRBuilder class.
+
+    Enables users to build up a Relay environment and program.
+    """
+
+    def __init__(self):
+        self.bindings = [{}]
+        self.scopes = [{}]
+        self.params = []
+        self.ret_values = [None]
+        self.env = Environment({})
+
+    def enter_scope(self, params=None):
+        if not params:
+            params = []
+
+        self.bindings.append({})
+        self.scopes.append({})
+        self.params.append(params)
+        self.ret_values.append(None)
+
+    def exit_scope(self):
+        bindings = self.bindings.pop()
+        scopes = self.scopes.pop()
+        params = self.params.pop()
+        ret_value = self.ret_values.pop()
+        return bindings, scopes, params, ret_value
+
+    #pylint: disable=invalid-name
+    def bind(self, name, value, ty):
+        lv = Var(name)
+        self.scopes[-1][name] = lv
+        self.bindings[-1][lv] = (value, ty)
+        return lv
```

Comment:
would be good to add doc for these functions.

Reply:
👍